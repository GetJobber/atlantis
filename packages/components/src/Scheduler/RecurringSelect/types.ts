import { DayOfMonth, DurationPeriod, WeekDay } from "../types";

export interface InitialStateMap {
  [DurationPeriod.Day]: {
    type: DurationPeriod.Day;
  };
  [DurationPeriod.Week]: {
    type: DurationPeriod.Week;
    weekDays: Set<WeekDay>;
  };
  [DurationPeriod.DayOfMonth]: {
    type: DurationPeriod.DayOfMonth;
    date: Set<DayOfMonth>;
  };
  [DurationPeriod.WeekDayOfMonth]: {
    type: DurationPeriod.WeekDayOfMonth;
    dayOfWeek: [Set<WeekDay>, Set<WeekDay>, Set<WeekDay>, Set<WeekDay>];
  };
  [DurationPeriod.Year]: {
    type: DurationPeriod.Year;
  };
}

export const typeInitialStateMap: InitialStateMap = {
  [DurationPeriod.Day]: {
    type: DurationPeriod.Day,
  },
  [DurationPeriod.Week]: {
    type: DurationPeriod.Week,
    weekDays: new Set<WeekDay>(),
  },
  [DurationPeriod.DayOfMonth]: {
    type: DurationPeriod.DayOfMonth,
    date: new Set<DayOfMonth>(),
  },
  [DurationPeriod.WeekDayOfMonth]: {
    type: DurationPeriod.WeekDayOfMonth,
    dayOfWeek: [
      new Set<WeekDay>(),
      new Set<WeekDay>(),
      new Set<WeekDay>(),
      new Set<WeekDay>(),
    ],
  },
  [DurationPeriod.Year]: {
    type: DurationPeriod.Year,
  },
};
