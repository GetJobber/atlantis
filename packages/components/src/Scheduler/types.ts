import { CivilTime } from "@std-proposal/temporal";

export enum WeekDay {
  Sunday,
  Monday,
  Tuesday,
  Wednesday,
  Thursday,
  Friday,
  Saturday,
}
export type DayOfMonth =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | "LAST";

export enum DurationPeriod {
  Day = "Day",
  Week = "Week",
  DayOfMonth = "DayOfMonth",
  WeekDayOfMonth = "WeekDayOfMonth",
  Year = "Year",
  Visit = "Visit",
  AsNeeded = "AsNeeded",
}

export enum RecurrenceOptions {
  Daily,
  Weekly,
  BiWeekly,
  AsNeeded,
  Custom,
  DoesNotRepeat,
}

export interface ScheduleEnd {
  type: "duration" | "date";
  numOfPeriods: number;
  durationPeriod: DurationPeriod;
  date: Date;
}

export interface RecurrenceRuleDay {
  type: DurationPeriod.Day;
}

export interface RecurrenceRuleWeek {
  type: DurationPeriod.Week;
  weekDays: Set<WeekDay>;
}

export interface RecurrenceRuleYear {
  type: DurationPeriod.Year;
}

export interface RecurrenceRuleAsNeeded {
  type: DurationPeriod.AsNeeded;
}

export interface RecurrenceRuleDayOfMonth {
  type: DurationPeriod.DayOfMonth;
  date: Set<DayOfMonth>;
}

export interface RecurrenceRuleWeekDayOfMonth {
  type: DurationPeriod.WeekDayOfMonth;
  dayOfWeek: [Set<WeekDay>, Set<WeekDay>, Set<WeekDay>, Set<WeekDay>];
}

export type RecurrenceRule = {
  interval: number;
} & (
  | RecurrenceRuleDay
  | RecurrenceRuleWeek
  | RecurrenceRuleYear
  | RecurrenceRuleDayOfMonth
  | RecurrenceRuleWeekDayOfMonth
  | RecurrenceRuleAsNeeded
);

export interface Recurrence {
  rule: RecurrenceRule;
  ends: ScheduleEnd;
}

export interface SchedulerState {
  startDate: Date;
  startTime: CivilTime | undefined;
  endTime: CivilTime | undefined;
  recurrence: Recurrence;
}
