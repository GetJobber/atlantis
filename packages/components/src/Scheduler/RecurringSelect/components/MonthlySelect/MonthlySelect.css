
.radioContainer div[role="radiogroup"]{
    flex-direction: row;
}

.radioContainer label{
    margin-right: var(--space-base);
}
