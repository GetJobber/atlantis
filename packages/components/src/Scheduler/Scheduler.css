.title {
    display: flex;
}

.title > *:first-child {
    flex: 1 1 auto;
}

.datePicker sg-icon[icon="calendar"] {
    color: var(--color-greyBlue);
}

.container input:focus, .container select:focus {
  box-shadow: var(--shadow-focus);
}

.disabled h5 {
color: var(--color-disabled);
}
