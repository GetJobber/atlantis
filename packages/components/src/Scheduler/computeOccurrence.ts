/**
 * The exported computeOccurrence function is used to determine complex scheduling of jobs, invoicing, etc.
 * Given on object of type Recurrence, and a start date, it will return an object with the following:
 *   firstOccurrence: the date of the first occurrence
 *   lastOccurrence: the date of the last occurrence
 *   totalOccurrences: the number of dates to be scheduled
 *   scheduledDates: an array containing the dates where an occurrence is to be scheduled
 */

import RRule, { Frequency, Weekday } from "rrule";
import {
  DayOfMonth,
  DurationPeriod,
  Recurrence,
  RecurrenceRule,
  ScheduleEnd,
  WeekDay,
} from "./types";

const NO_SCHEDULED_DATE = "NO_SCHEDULED_DATE";

/**
 * RRule treats monday as the 0th day, JS treats sunday as the 0th day.
 * This array allows for easy mapping between RRule and JS.
 */
const rruleWeekDays = [
  RRule.SU,
  RRule.MO,
  RRule.TU,
  RRule.WE,
  RRule.TH,
  RRule.FR,
  RRule.SA,
];

const durationPeriodToRRuleFreq = (durationPeriod: DurationPeriod) => {
  switch (durationPeriod) {
    case DurationPeriod.Day:
      return RRule.DAILY;
    case DurationPeriod.Week:
      return RRule.WEEKLY;
    case DurationPeriod.DayOfMonth:
      return RRule.MONTHLY;
    case DurationPeriod.WeekDayOfMonth:
      return RRule.MONTHLY;
    case DurationPeriod.Year:
      return RRule.YEARLY;
    default:
      return undefined;
  }
};

// This is a weird function - let me explain. Rrule needs to take UTC dates in to work correctly.
// If we just converted to Date.UTC(year, month, day) there are cases where converting to UTC
// will change the day (and possibly month). Take for example, Nov.1, 2021 in Edmonton TZ. Converting to UTC
// will return a date of Oct.31, 2021, because we don't factor in time at all - these dates are at 00:00.
// Adding the TZ offset will give you a date in UTC that is technically +/- the actual time in UTC, however it is in UTC
// so it doesn't break rrule. This ensures our rrules don't break on edge cases such as timezone changes, different timezones, etc.
// It is essentially a UTC date (wrt year, month, day) framed from the reference of the users locale.
export const convertToUtcDate = (date: Date) => {
  return new Date(
    Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()) +
      new Date().getTimezoneOffset() * 60 * 1000,
  );
};

const computeEndDateOrCount = (
  startDate: Date,
  recurrenceRule: RecurrenceRule,
  duration: ScheduleEnd,
  frequency: Frequency,
) => {
  if (duration.durationPeriod === DurationPeriod.Visit) {
    return { count: duration.numOfPeriods };
  }

  const startToFinishDates = new RRule({
    dtstart: startDate,
    freq: durationPeriodToRRuleFreq(duration.durationPeriod),
    count: duration.numOfPeriods + 1,
  });

  const rrule = new RRule({
    dtstart: startDate,
    freq: frequency,
    interval: recurrenceRule.interval,
    until: startToFinishDates.all()[startToFinishDates.all().length - 1],
  });

  return { until: rrule.all()[rrule.all().length - 1] };
};

const getEndDateOrCount = (
  startDate: Date,
  rule: RecurrenceRule,
  ends: ScheduleEnd,
  frequency: Frequency,
) => {
  return ends.type === "date"
    ? { until: convertToUtcDate(ends.date) }
    : computeEndDateOrCount(startDate, rule, ends, frequency);
};

const getWeekDayArray = (weekDays: Set<WeekDay>) => {
  const weekDayArray: Weekday[] = [];
  weekDays.forEach((weekDay: number) => {
    weekDayArray.push(rruleWeekDays[weekDay]);
  });

  return weekDayArray;
};

const getMonthDayArray = (daysOfMonth: Set<DayOfMonth>) => {
  const daysOfMonthArray: number[] = [];
  daysOfMonth.forEach((day: DayOfMonth) => {
    if (day === "LAST") {
      daysOfMonthArray.push(-1);
    } else {
      daysOfMonthArray.push(day);
    }
  });

  return daysOfMonthArray;
};

const getWeekDayOfMonthArray = (daysOfWeeks: Set<WeekDay>[]) => {
  const weekDayOfMonthArray: Weekday[] = [];
  daysOfWeeks.forEach((weekDaySet: Set<WeekDay>, index: number) => {
    weekDaySet.forEach((weekDay: number) => {
      weekDayOfMonthArray.push(rruleWeekDays[weekDay].nth(index + 1));
    });
  });

  return weekDayOfMonthArray;
};

const daySelectionIsEmpty = (selectedDays: number[] | Weekday[]) => {
  return selectedDays.length < 1;
};

const getOccurrenceResult = (rrule: RRule) => {
  return {
    firstOccurrence:
      rrule.all().length > 0 ? rrule.all()[0] : NO_SCHEDULED_DATE,
    lastOccurrence:
      rrule.all().length > 0
        ? rrule.all()[rrule.all().length - 1]
        : NO_SCHEDULED_DATE,
    totalOccurrence: rrule.all().length,
    scheduledDates: rrule.all(),
    summaryString: rrule.toText(),
  };
};

const getEmptyOccurrenceResult = () => {
  return {
    firstOccurrence: NO_SCHEDULED_DATE,
    lastOccurrence: NO_SCHEDULED_DATE,
    totalOccurrence: 0,
    scheduledDates: [],
    summaryString: "",
  };
};

const computeSingleOccurrence = (startDate: Date) => {
  return {
    firstOccurrence: startDate,
    lastOccurrence: undefined,
    totalOccurrence: 1,
    scheduledDates: [startDate],
    summaryString: undefined,
  };
};

const computeNDayRepeat = (
  startDate: Date,
  rule: RecurrenceRule,
  ends: ScheduleEnd,
) => {
  const endDateOrCount = getEndDateOrCount(startDate, rule, ends, RRule.DAILY);

  const rrule = new RRule({
    freq: RRule.DAILY,
    interval: rule.interval,
    dtstart: startDate,
    ...endDateOrCount,
  });

  return getOccurrenceResult(rrule);
};

const computeNWeekRepeat = (
  startDate: Date,
  rule: RecurrenceRule,
  ends: ScheduleEnd,
  weekDays: Set<WeekDay>,
) => {
  const endDateOrCount = getEndDateOrCount(startDate, rule, ends, RRule.WEEKLY);
  const weekDayArray = getWeekDayArray(weekDays);

  if (daySelectionIsEmpty(weekDayArray)) {
    return getEmptyOccurrenceResult();
  }

  const rrule = new RRule({
    freq: RRule.WEEKLY,
    interval: rule.interval,
    dtstart: startDate,
    byweekday: weekDayArray,
    ...endDateOrCount,
  });

  return getOccurrenceResult(rrule);
};

const computeDayOfMonthRepeat = (
  startDate: Date,
  rule: RecurrenceRule,
  ends: ScheduleEnd,
  daysOfMonth: Set<DayOfMonth>,
) => {
  const endDateOrCount = getEndDateOrCount(
    startDate,
    rule,
    ends,
    RRule.MONTHLY,
  );
  const monthDayArray = getMonthDayArray(daysOfMonth);

  if (daySelectionIsEmpty(monthDayArray)) {
    return getEmptyOccurrenceResult();
  }

  const rrule = new RRule({
    freq: RRule.MONTHLY,
    interval: rule.interval,
    dtstart: startDate,
    bymonthday: monthDayArray,
    ...endDateOrCount,
  });

  return getOccurrenceResult(rrule);
};

const computeWeekDayOfMonthRepeat = (
  startDate: Date,
  rule: RecurrenceRule,
  ends: ScheduleEnd,
  daysOfWeeks: Set<WeekDay>[],
) => {
  const endDateOrCount = getEndDateOrCount(
    startDate,
    rule,
    ends,
    RRule.MONTHLY,
  );
  const weekDayOfMonthArray = getWeekDayOfMonthArray(daysOfWeeks);

  if (daySelectionIsEmpty(weekDayOfMonthArray)) {
    return getEmptyOccurrenceResult();
  }

  const rrule = new RRule({
    freq: RRule.MONTHLY,
    interval: rule.interval,
    dtstart: startDate,
    byweekday: weekDayOfMonthArray,
    ...endDateOrCount,
  });

  return getOccurrenceResult(rrule);
};

const computeNYearRepeat = (
  startDate: Date,
  rule: RecurrenceRule,
  ends: ScheduleEnd,
) => {
  const endDateOrCount = getEndDateOrCount(startDate, rule, ends, RRule.YEARLY);

  const rrule = new RRule({
    freq: RRule.YEARLY,
    interval: rule.interval,
    dtstart: startDate,
    ...endDateOrCount,
  });

  return getOccurrenceResult(rrule);
};

const computeRecurringSchedule = (
  startDate: Date,
  recurrence: Exclude<Recurrence, { recurs: false }>,
) => {
  switch (recurrence.rule.type) {
    case DurationPeriod.Day:
      return computeNDayRepeat(startDate, recurrence.rule, recurrence.ends);
    case DurationPeriod.Week:
      return computeNWeekRepeat(
        startDate,
        recurrence.rule,
        recurrence.ends,
        recurrence.rule.weekDays,
      );
    case DurationPeriod.DayOfMonth:
      return computeDayOfMonthRepeat(
        startDate,
        recurrence.rule,
        recurrence.ends,
        recurrence.rule.date,
      );
    case DurationPeriod.WeekDayOfMonth:
      return computeWeekDayOfMonthRepeat(
        startDate,
        recurrence.rule,
        recurrence.ends,
        recurrence.rule.dayOfWeek,
      );
    case DurationPeriod.Year:
      return computeNYearRepeat(startDate, recurrence.rule, recurrence.ends);
    default:
      return computeSingleOccurrence(startDate);
  }
};

export const computeOccurrence = (startDate: Date, recurrence: Recurrence) => {
  const utcStartDate = convertToUtcDate(startDate);
  return computeRecurringSchedule(utcStartDate, recurrence);
};
