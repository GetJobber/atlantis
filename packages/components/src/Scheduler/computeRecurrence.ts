import {
  DurationPeriod,
  Recurrence,
  RecurrenceOptions,
  ScheduleEnd,
} from "./types";

export const computeRecurrence = (
  startDate: Date,
  from: Recurrence,
  preset: Exclude<
    RecurrenceOptions,
    RecurrenceOptions.AsNeeded // | RecurrenceOptions.DoesNotRepeat
  >,
): Recurrence => {
  const ends: ScheduleEnd = from.ends
    ? from.ends
    : {
        type: "date",
        date: new Date(),
        durationPeriod: DurationPeriod.Week,
        numOfPeriods: 6,
      };

  // tslint:disable-next-line: switch-default
  switch (preset) {
    case RecurrenceOptions.DoesNotRepeat: {
      return {
        rule: from.rule,
        ends,
      };
    }
    case RecurrenceOptions.Daily: {
      return {
        rule: {
          type: DurationPeriod.Day,
          interval: 1,
        },
        ends,
      };
    }

    case RecurrenceOptions.Weekly: {
      return {
        rule: {
          type: DurationPeriod.Week,
          interval: 1,
          weekDays: new Set([startDate.getDay()]),
        },
        ends,
      };
    }

    case RecurrenceOptions.BiWeekly: {
      return {
        rule: {
          type: DurationPeriod.Week,
          interval: 2,
          weekDays: new Set([startDate.getDay()]),
        },
        ends,
      };
    }

    case RecurrenceOptions.Custom: {
      return {
        rule: from.rule,
        ends,
      };
    }
  }
};
