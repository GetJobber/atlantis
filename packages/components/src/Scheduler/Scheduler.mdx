---
name: Scheduler
menu: Components
route: /components/scheduler
showDirectoryLink: true
---

import { Playground, Props } from "docz";
import { ComponentStatus } from "@jobber/docx";
import { Scheduler } from ".";
import { Content } from "@jobber/components/Content";
import { Text } from "@jobber/components/Text";
import { Checkbox } from "@jobber/components/Checkbox";
import { Divider } from "@jobber/components/Divider";
import { Avatar } from "@jobber/components/Avatar";
import { useState } from "react";

# Scheduler

<ComponentStatus stage="ready" responsive="yes" accessible="yes" />

Scheduler, fill me out

```ts
import { Scheduler } from "@jobber/components/Scheduler";
```

<Playground>
  <Scheduler />
</Playground>
