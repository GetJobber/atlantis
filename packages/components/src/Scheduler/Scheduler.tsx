import React, { useMemo } from "react";
import { CivilTime } from "@std-proposal/temporal";
import classNames from "classnames";
import { RecurringSelect } from "./RecurringSelect";
import {
  Recurrence,
  RecurrenceOptions,
  RecurrenceRule,
  ScheduleEnd,
  SchedulerState,
} from "./types";
import styles from "./Scheduler.css";
import { computeRecurrence } from "./computeRecurrence";
import { ScheduleSummary } from "./ScheduleSummary";
import { computeOccurrence } from "./computeOccurrence";
import { strFormatDate } from "../FormatDate";
import { Content } from "../Content";
import { Card } from "../Card";

export interface SchedulerProps {
  startDate: Date;
  startTime: CivilTime | undefined;
  endTime: CivilTime | undefined;
  recurrence: Recurrence;
  selectedRecurringOption: RecurrenceOptions;
  summaryString?: string;
  onChange(state: Partial<SchedulerState>): void;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line max-statements
export const Scheduler = (props: SchedulerProps) => {
  const getRecurrenceRule = (option: RecurrenceOptions) => {
    if (option === RecurrenceOptions.AsNeeded) {
      return props.recurrence;
    }

    return computeRecurrence(props.startDate, props.recurrence, option);
  };

  const onRecurrenceChange = (newRecurrence: {
    ends: ScheduleEnd;
    rule: RecurrenceRule;
  }) =>
    props.onChange({
      recurrence: {
        ...newRecurrence,
      },
    });

  const schedule = useMemo(
    () => computeOccurrence(props.startDate, props.recurrence),
    [props.startDate, props.recurrence],
  );

  const containerClass = classNames(styles.container);

  const showAdditionalDetail =
    (props.recurrence.ends.type === "duration" ||
      (props.recurrence.ends.type === "date" &&
        props.recurrence.ends.date > props.startDate)) &&
    props.selectedRecurringOption !== RecurrenceOptions.AsNeeded &&
    schedule.lastOccurrence !== undefined;

  return (
    <div className={containerClass}>
      <Card>
        <Content>
          <>
            <Content spacing="small">
              <RecurringSelect
                recurrenceRule={props.recurrence.rule}
                recurrenceEnds={props.recurrence.ends}
                disabled={false}
                onRecurrenceChange={onRecurrenceChange}
              />
            </Content>
          </>
          {showAdditionalDetail && (
            <Content spacing="small">
              <ScheduleSummary
                disabled={false}
                startDate={
                  typeof schedule.firstOccurrence === "string"
                    ? "-"
                    : strFormatDate(schedule.firstOccurrence)
                }
                endDate={
                  schedule.lastOccurrence instanceof Date
                    ? strFormatDate(schedule.lastOccurrence)
                    : "-"
                }
                totalOccurences={schedule.totalOccurrence}
                summaryString={
                  schedule.summaryString
                    ? `Visits ${schedule.summaryString}`
                    : ""
                }
              />
            </Content>
          )}
        </Content>
      </Card>
    </div>
  );
};
