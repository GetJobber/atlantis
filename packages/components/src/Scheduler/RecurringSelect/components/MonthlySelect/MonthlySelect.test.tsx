/* tslint:disable no-relative-imports */
import React from "react";
import { cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import { MonthlySelect } from "./MonthlySelect";
// eslint-disable-next-line
import { DurationPeriod, WeekDay } from "../../../Scheduler/types";

afterEach(cleanup);

let onChange: jest.Mock;

describe("MonthlySelect", () => {
  beforeEach(() => {
    onChange = jest.fn();
  });

  it("should render the component", () => {
    const tree = renderer
      .create(
        <MonthlySelect
          disabled={false}
          onChange={onChange}
          selectedMonthOption={DurationPeriod.DayOfMonth}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
