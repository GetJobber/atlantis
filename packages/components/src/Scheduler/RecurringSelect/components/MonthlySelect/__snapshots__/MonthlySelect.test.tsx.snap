// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlySelect should render the component 1`] = `
<div
  className="radioContainer"
>
  <div
    aria-label="Monthly Frequency Selector"
    className="_21aQKdbfCYC-03CNAUhBXW"
    role="radiogroup"
  >
    <div>
      <input
        checked={true}
        className="_11AeV22s7aou-ZpJgQCiPD"
        disabled={false}
        id="DayOfMonth_123e4567-e89b-12d3-a456-426655440002"
        name="123e4567-e89b-12d3-a456-426655440001"
        onChange={[Function]}
        type="radio"
        value="DayOfMonth"
      />
      <label
        className="IiW-lj1gmbjv30FHatRaU"
        htmlFor="DayOfMonth_123e4567-e89b-12d3-a456-426655440002"
      >
        Day of month
      </label>
    </div>
    <div>
      <input
        checked={false}
        className="_11AeV22s7aou-ZpJgQCiPD"
        disabled={false}
        id="WeekDayOfMonth_123e4567-e89b-12d3-a456-426655440003"
        name="123e4567-e89b-12d3-a456-426655440001"
        onChange={[Function]}
        type="radio"
        value="WeekDayOfMonth"
      />
      <label
        className="IiW-lj1gmbjv30FHatRaU"
        htmlFor="WeekDayOfMonth_123e4567-e89b-12d3-a456-426655440003"
      >
        Day of week
      </label>
    </div>
  </div>
</div>
`;
